EchoPro/survey.py

```diff
@@ -944,10 +840,287 @@ def strata_age_binned_weight_proportions( self ,
             }
         } )
 
-    # @staticmethod
-    # def initialize_kriging_mesh():
+    #!!! TODO : Provide argument that will exclude age-0 and age-1 fish when flagged
+    def nasc_to_biomass_conversion( self , 
+                                    species_id: np.float64  ):
+        """
+        Converts integrated acoustic backscatter (NASC) into estimates of 
+        areal number/biomass densities, total abundance, and total biomass
+
+        Parameters
+        ----------
+        species_id : np.float64
+            Numeric code representing a particular species of interest
+
+        Notes
+        -----
+        This function converts NASC into estimates of population-level metrics 
+        (abundance, biomass, areal densities) stratified by transects, sex, 
+        length-based strata, and age.
+        """ 
+        
+        ### Prepare initial dataframes used for calculation population statistics
+        # Construct georeferenced dataframe containing NASC data
+        nasc_df = self.acoustics[ 'nasc' ][ 'nasc_df' ]        
+
+        ### Calculate transect interval lengths and areas, plus impute/replace possible 'erroneous' 
+        ### values at the end of each transect line
+        nasc_interval_df = (
+            nasc_df
+            # Calculate along-transect interval distances
+            .pipe( lambda df: df.assign( interval = df[ 'vessel_log_start' ].diff( periods = -1 ).abs( ) ) 
+                                .replace( np.nan , df[ 'vessel_log_end' ].iloc[ -1 ] - df[ 'vessel_log_start' ].iloc[ -1 ] ) )
+            # Replace likely erroneous interval lengths associated with the edges of each transect
+            .pipe( lambda df: df.assign( median_interval = np.median( df[ 'interval' ] ) )
+                                .assign( interval = lambda x: np.where( np.abs( x[ 'interval' ] - x[ 'median_interval' ] > 0.05 ) ,
+                                                                        x.vessel_log_end - x.vessel_log_start ,
+                                                                        x.interval ) ) )
+            # Calculate interval area
+            .pipe( lambda df: df.assign( interval_area = df[ 'interval' ] * df[ 'transect_spacing' ] ) )                            
+            # Keep dataframe tidy by only retaining the necessary columns/variables
+            .loc[ : , [ 'latitude' , 'longitude' , 'transect_num' , 'stratum_num' , 'haul_num' , 'interval_area' , 'NASC_all_ages' , 'NASC_no_age1' ] ]
+        )
+
+        ### Call additional dataframes needed to merge with the NASC data and subsequently calculate
+        ### population-level metrics (and later statistics)
+        # Sex-stratum-indexed proportions and average weight
+        weight_sex_strata = (
+            self.biology[ 'weight' ][ 'weight_strata_df' ]
+            .copy()
+            .rename( columns = { 'stratum': 'stratum_num' } )
```

Comment:
I wonder if this renaming means that we could rename this column in the data loader.

Reply:
I think there are two ways to go about this: 

1. I could not be lazy and accidentally name the data column `stratum` instead of `stratum_num`.
2. We can drop the appended `_num` in the data loader so it is consistent with other variables (e.g. `length` instead of `length_num`). That may be a poor example since, as far as I understand, `*_num` indicates the numbered transect, haul, and stratum, and therefore acts as a flag for "hey, this is an indexing variable we care about".

It's not a particularly onerous task to type out '_num', and appending that text also is (probably) never the primary contributor toward very long lines of code that need to be better formatted/broken up. If `*_num` is, in fact, a flag (of sorts) of an ordinal numbered variable, then it may be useful to devise a data dictionary/glossary that at least defines its meaning. I think this also more generally applies to where we append `*_df`, `*_gdf`, `*_arr`, and other similar strings. It is fairly intuitive from our end, but it may not be for folks interacting with the `Survey` objects. This already exists to some degree here:

https://github.com/uw-echospace/EchoPro/blob/f653ba74f445ad97c365bfc077bde602a3b8ab8b/docs/classes-datastructures.md?plain=1#L25